lightning-invoice/src/utils.rs

```diff
@@ -21,6 +21,99 @@ use core::convert::TryInto;
 use core::ops::Deref;
 use core::time::Duration;
 
+#[cfg(feature = "std")]
+/// Utility to create an invoice that can be paid to one of multiple nodes, or a "phantom invoice."
+/// See [`PhantomKeysManager`] for more information on phantom node payments.
+///
+/// `phantom_route_hints` parameter:
+/// * Contains channel info for all nodes participating in the phantom invoice
+/// * Entries are retrieved from a call to [`ChannelManager::get_phantom_route_hints`] on each
+///   participating node
+/// * It is fine to cache `phantom_route_hints` and reuse it across invoices, as long as the data is
+///   updated when a channel becomes disabled or closes
+/// * Note that if too many channels are included in the [`PhantomRouteHints`]s' `channels` fields,
+///   the invoice may be too long for QR code scanning. To fix this, `channels` field(s) may be pared
+///   down
+///
+/// `payment_hash` and `payment_secret` come from [`ChannelManager::create_inbound_payment`] or
+/// [`ChannelManager::create_inbound_payment_for_hash`]. These values can be retrieved from any
+/// participating node.
+///
+/// Note that the provided `keys_manager`'s `KeysInterface` implementation must support phantom
+/// invoices in its `sign_invoice` implementation ([`PhantomKeysManager`] satisfies this
+/// requirement).
+///
+/// [`PhantomKeysManager`]: lightning::chain::keysinterface::PhantomKeysManager
+/// [`ChannelManager::get_phantom_route_hints`]: lightning::ln::channelmanager::ChannelManager::get_phantom_route_hints
+/// [`PhantomRouteHints`]: lightning::ln::channelmanager::PhantomRouteHints
+pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
+	amt_msat: Option<u64>, description: String, payment_hash: PaymentHash, payment_secret:
+	PaymentSecret, phantom_route_hints: &Vec<PhantomRouteHints>, keys_manager: K, network: Currency
```

Comment:
Bindings can't map `&Vec<>` as vecs have to be mapped to something C-readable, there's no good way to map a reference to a vec without the whole vec being opaque. Bindings can really only map `Vec<Object>` (if we want move semantics) or `&[&Object]` (if we want reference semantics. I recently added support for `&[Object]` but it'd be nice to avoid using that too much as it has to clone each element as we cross from C to Rust or back.